### ISSUE TITLE
Build dashboard sign in silently is failing

### ISSUE THREAD

**Issue by CaseyHillers** [flutter/flutter#121275]
This prevents users from being able to run API requests (such as rerunning jobs).

As a workaround, I can manually click sign in, and go through the auth flow when I need to authenticate. Instead, we should fix the silent sign in so users are automatically signed in.

**Comment by CaseyHillers** [comment]
\cc @nehalvpatel

**Comment by CaseyHillers** [comment]
This might be related to the recent changes to google_sign_in_web (see https://pub.dev/packages/google_sign_in_web for the migration guide). There's nothing obvious to me from there that we need to migrate.

**Comment by nehalvpatel** [comment]
Hm, it looks like `signInSilently` which we call on load doesn't authorize scopes anymore. But then the Sign In button performs the authorization and finally loads the full user.

cc @ditman from this PR: https://github.com/flutter/plugins/pull/6921

Does this mean that we'll have to call `requestScopes` after `signInSilently`? I'm not sure if `onCurrentUserChanged` is still expected to fire after `signInSilently`.

https://github.com/flutter/cocoon/blob/8756bd3c825271dc1443cb58c8669cc590daa71f/dashboard/lib/service/google_authentication.dart#L26

**Comment by ditman** [comment]
Google Sign In SDK has undergone **major** changes (hence the major version bump of the package).

I've documented all the major changes in behavior, and what to do to overcome/workaround/rework some of those, here:

* https://pub.dev/packages/google_sign_in_web#differences-between-google-identity-services-sdk-and-google-sign-in-for-web-sdk

TL;DR, the things that are affecting Cocoon the most:

* The GIS SDK no longer handles sign-in state and user sessions, it only provides Authentication credentials for the moment the user did authenticate. This means: refresh? login again.
* The GIS SDK no longer is able to renew Authorization sessions on the web. Once the token expires, API requests will begin to fail with "unauthorized" error codes, and user Authorization is required again.

I think that for cocoon this means that:

* Displaying the dashboard should **not** require user authentication (so it can be left unattended, on a screen, with it constantly updating)
* Interacting with the dashboard may now *fail* with an unauthorized request error, because neither authentication nor authorization is auto-renewed.
  * In this case, the user should be prompted to `signIn` again so the correct `scopes` are granted again. Authorization tokens last for 1h, and cannot be extended from the client-side. See [token handling](https://developers.google.com/identity/oauth2/web/guides/migration-to-gis#token_handling).

We've created an issue with the GIS team to at least return an `idToken` from the calls to renew "scopes" at least, but I understand the frustration, the changes to how google provides authentication and authorization are major :/

(Feel free to add me to whatever situation rooms you have about this)

**Comment by ditman** [comment]
The way we've solved this in other apps is:

* signInSilently always returns null [(Here's why)](https://pub.dev/packages/google_sign_in_web#signinsilently-always-returns-null), but if it can authenticate you through the One Tap UX, it'll keep your `idToken` around.
* signIn will now be able to authorize (and sign-in if needed) the use, and that's when the `onCurrentUserChanged` fires.

Use `signInSilently` when building every page of the site, then render the `signIn` button as needed. This will authorize the required scopes.

Apps should start distinguishing between two states: "user signed in" and "scopes authorized". (This is because in the medium term, authentication may come from a different provider (FedCM, FIDO2, other authentication providers) than authorization (Google, other API providers...))

**Comment by ricardoamador** [comment]
@ditman do you expect the changes that need to be made here are particularly involved or is this something fairly simple to fix?

**Comment by ditman** [comment]
@ricardoamador I don't know how your backend works exactly, but we can have a meeting and discuss!

**Comment by ricardoamador** [comment]
@ditman cool I can set something up for Thursday or Friday. I appreciate the help.

**Comment by ditman** [comment]
@ricardoamador seen your invite, talk to you on Thursday!

**Comment by ricardoamador** [comment]
Met with David today and he provided some good info and is working on helping us support a new login method. He is going to make some changes in flutter that will allow us to use the google sign-in button with some modification.
https://github.com/flutter/flutter/issues/121781

**Comment by keyonghan** [comment]
Based on the above communications, assigning to @ricardoamador to follow up. Moving out of ticket queue as this doesn't seem a trivial fix.

**Comment by ricardoamador** [comment]
No problem. I will follow up with David.

**Comment by ricardoamador** [comment]
Okay how would I add this change to our project for testing?

**Comment by ditman** [comment]
@ricardoamador we could override ~4 packages from a branch I yet don't have, but it's better if you let me start landing them :P

**Comment by ricardoamador** [comment]
@ditman sure no problem. let me know when you are ready. Let's go with whichever method is easiest.

**Comment by ditman** [comment]
I have pushed all I have to a branch. I made some progress in the configurability of the button. I'll start validating the approach next week (I need some more tests, and to make sure that I'm not breaking Mobile with my changes!)

**Comment by ditman** [comment]
Latest update:

* I have now a PR: https://github.com/flutter/packages/pull/3461 (draft)
* The configurability of the button is pretty much done.
* Tests are written for 2/3 of the packages (need tests for the _web package)
* _(I am now 100% sure that my example is breaking Mobile, so I'll need to update the example to support the cross-platform use case 😅)_

**Comment by ditman** [comment]
Latest update

* PR is still the same https://github.com/flutter/packages/pull/3461
* Configurability of the button is done and tested.
* Mobile example fixed (needed a conditional import)
* Tests are 90% done. This will go into proper PR today.

**Comment by ricardoamador** [comment]
That's great new David! Thank you for doing this!

**Comment by ditman** [comment]
A fix for this is coming. I have posted some quick migration instructions here:

* https://github.com/flutter/flutter/issues/124297

I think once google_sign_in: 6.1.0 lands, we'll be able to fix cocoon in a matter of minutes.

**Comment by ricardoamador** [comment]
Hey that is great news @ditman! Looking forward to it.

**Comment by ditman** [comment]
`google_sign_in: ^6.1.0` has landed, I'll try to get cocoon going locally to try to update the sign-in!

* https://pub.dev/packages/google_sign_in/changelog#610

**Comment by ditman** [comment]
Here you go peeps!

https://github.com/flutter/cocoon/pull/2601

I tested the web locally with fake data, not sure if you guys stage this in any way to take a look with actual data.
